src/gd_heif.c

```diff
@@ -0,0 +1,855 @@
+
+    You can inspect the sx and sy members of the image to determine
+    its size. The image must eventually be destroyed using
+    <gdImageDestroy>.
+
+    *The returned image is always a truecolor image.*
+
+  Variants:
+
+    <gdImageCreateFromHeif> AVIF method counterpart.
+
+  Parameters:
+
+    infile - The input FILE pointer.
+
+  Returns:
+
+    A pointer to the new *truecolor* image.  This will need to be
+    destroyed with <gdImageDestroy> once it is no longer needed.
+
+    On error, returns NULL.
+*/
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvif(FILE *inFile)
+{
+	gdImagePtr im;
+	gdIOCtx *in = gdNewFileCtx(inFile);
+	if (!in) {
+		return 0;
+	}
+	im = _gdImageCreateFromHeifCtx(in, GD_HEIF_BRAND_AVIF | GD_HEIF_BRAND_MIF1);
+	in->gd_free(in);
+
+	return im;
+}
+
+/*
+  Function: gdImageCreateFromAvifPtr
+
+    See <gdImageCreateFromAvif>.
+
+  Parameters:
+
+    size            - size of AVIF data in bytes.
+    data            - pointer to AVIF data.
+*/
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvifPtr(int size, void *data)
+{
+	gdImagePtr im;
+	gdIOCtx *in = gdNewDynamicCtxEx(size, data, 0);
+	if (!in)
+		return 0;
+	im = _gdImageCreateFromHeifCtx(in, GD_HEIF_BRAND_AVIF | GD_HEIF_BRAND_MIF1);
+	in->gd_free(in);
+	return im;
+}
+
+/*
+  Function: gdImageCreateFromAvifCtx
+
+    See <gdImageCreateFromAvif>.
+*/
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvifCtx(gdIOCtx *infile)
+{
+	return _gdImageCreateFromHeifCtx(infile, GD_HEIF_BRAND_AVIF | GD_HEIF_BRAND_MIF1);
+}
+
+
+static struct heif_error _gdImageWriteHeif(struct heif_context *heif_ctx, const void *data, size_t size, void *userdata)
+{
+	gdIOCtx *outfile;
+	struct heif_error err;
+
+	outfile = (gdIOCtx *)userdata;
+
+	gdPutBuf(data, size, outfile);
+
+	err.code = heif_error_Ok;
+	err.subcode = heif_suberror_Unspecified;
+	err.message = "";
+
+	return err;
+}
+
+/* returns GD_TRUE on success, GD_FALSE on failure */
+static int _gdImageHeifCtx(gdImagePtr im, gdIOCtx *outfile, int quality, gdHeifCodec codec, gdHeifChroma chroma)
+{
+	struct heif_context *heif_ctx;
+	struct heif_encoder *heif_enc;
+	struct heif_image *heif_im;
+	struct heif_writer heif_wr;
+	struct heif_error err;
+	uint8_t *rgba;
+	int x, y;
+	uint8_t *p;
+
+	if (im == NULL) {
+		return GD_FALSE;
+	}
+
+	if (codec != GD_HEIF_CODEC_HEVC && codec != GD_HEIF_CODEC_AV1) {
+		gd_error("Unsupported format by heif");
+		return GD_FALSE;
+	}
+
+	if (!gdImageTrueColor(im)) {
+		gd_error("Palette image not supported by heif\n");
+		return GD_FALSE;
+	}
+
+	if (overflow2(gdImageSX(im), 4)) {
+		return GD_FALSE;
+	}
+
+	if (overflow2(gdImageSX(im) * 4, gdImageSY(im))) {
+		return GD_FALSE;
+	}
+
+	heif_ctx = heif_context_alloc();
+	if (heif_ctx == NULL) {
+		gd_error("gd-heif could not allocate context\n");
+		return NULL;
+	}
+	err = heif_context_get_encoder_for_format(heif_ctx, codec, &heif_enc);
+	if (err.code != heif_error_Ok) {
+		gd_error("gd-heif image creation failed\n");
```

Comment:
since this can fail if the codec is not supported in the host env, how about tweaking the message like:

 gd_error("gd-heif image creation failed (missing codec support?)\n");